rts-core/src/hooks.rs

```diff
@@ -0,0 +1,76 @@
+use pyo3::prelude::*;
+use pyo3::types::IntoPyDict;
+
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::sync::{Arc, Mutex};
+
+use crate::entity::game_actions::Action;
+use crate::exceptions::RtsException;
+
+type InnerMap = Arc<Mutex<HashMap<u64, PyModule>>>;
+
+pub struct PythonCodeExecutor {
+    compiled_ai_code: InnerMap,
+}
+
+impl Default for PythonCodeExecutor {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PythonCodeExecutor {
+    pub fn register_new_ai(&self, ai_code: String) -> Result<(), RtsException> {
+        let hash = calculate_hash(&ai_code);
+
+        let compiled_ai_code_ptr = Arc::clone(&self.compiled_ai_code);
+        let mut compiled_ai_code_mutex = compiled_ai_code_ptr.lock().map_err(|_| {
+            RtsException::PythonException(format!(
+                "Failed to acquire mutex when register new ai with hash {}",
+                hash.clone()
+            ))
+        })?;
+
+        Python::with_gil(|py| {
+            let module = compile_python_code(py, &ai_code, &hash)
```

Comment:
I have an issue their. `compile_python_code()` return a `&PyModule`. I'd like to store it the hash map, but it requires the owner ship. I could solve this by using a lifetime, but well, I don't wanna :smile: